File: cpp_prac/kakaofilesort.cpp
#include <string>
#include <vector>
#include <cctype>
#include <algorithm>

using namespace std;

bool classcomp(pair<string,int> lhs, pair<string,int> rhs)
{
    string lhead,rhead;
    string lnum="00000",rnum="00000";
    int numst,numfi;
    for(numst=0; numst<lhs.first.size(); ++numst)
    {
        if(!isdigit(lhs.first[numst])) lhead.push_back(tolower(lhs.first[numst]));
        else break;
    }
    for(numfi=numst+1; numfi<lhs.first.size(); ++numfi){if(!isdigit(lhs.first[numfi])) break;}
    for(int i=5-numfi+numst; i<5; ++i){lnum[i]=lhs.first[numst++];}
    
    for(numst=0; numst<rhs.first.size(); ++numst)
    {
        if(!isdigit(rhs.first[numst])) rhead.push_back(tolower(rhs.first[numst]));
        else break;
    }
    for(numfi=numst+1; numfi<rhs.first.size(); ++numfi){if(!isdigit(rhs.first[numfi])) break;}
    for(int i=5-numfi+numst; i<5; ++i){rnum[i]=rhs.first[numst++];}
    
    if(lhead<rhead) return true;
    else if(lhead==rhead)
    {
        if(lnum<rnum) return true;
        else if(lnum==rnum)
        {
            if(lhs.second<rhs.second) return true;
        }
    }
    return false;
}

vector<string> solution(vector<string> files) {
    vector<pair<string,int>> answer;
    for(int i=0; i<files.size(); ++i){answer.push_back(make_pair(files[i],i));}
    sort(answer.begin(),answer.end(),classcomp);
    for(int i=0; i<answer.size(); ++i){files[i]=answer[i].first;}
    return files;
}

File: java_prac/kakaofilesort.java
import java.util.Collections;
import java.util.ArrayList;
import java.util.List;

public class strnum implements Comparable<strnum>
{
    public String str;
    public int num;
    public strnum(){str=""; num=0;}
    public strnum(String str, int num)
    {
        this.str=str;
        this.num=num;
    }
    public int compareTo(strnum rhs)
    {
        String lhead="",rhead="";
        String lnum="00000",rnum="00000";
        int numst,numfi;
        for(numst=0; numst<this.str.length(); ++numst)
        {
            if(Character.isDigit(this.str.charAt(numst)))
            {
                lhead=this.str.substring(0,numst).toLowerCase();
                break;
            }
        }
        for(numfi=numst+1; numfi<this.str.length(); ++numfi){if(!Character.isDigit(this.str.charAt(numfi))) break;}
        lnum=lnum.substring(0,5-numfi+numst)+this.str.substring(numst,numfi);
           
        for(numst=0; numst<rhs.str.length(); ++numst)
        {
            if(Character.isDigit(rhs.str.charAt(numst)))
            {
                rhead=rhs.str.substring(0,numst).toLowerCase();
                break;
            }
        }
        for(numfi=numst+1; numfi<rhs.str.length(); ++numfi){if(!Character.isDigit(rhs.str.charAt(numfi))) break;}
        rnum=rnum.substring(0,5-numfi+numst)+rhs.str.substring(numst,numfi);
        if(lhead.compareTo(rhead)==1) return 1;
        else if(lhead.compareTo(rhead)==0)
        {
            if(lnum.compareTo(rnum)==1) return 1;
            else if(lnum.compareTo(rnum)==0)
            {
                if(this.num>rhs.num) return 1;
            }
        }
        return -1;
    }
    
    
}
class Solution {
  public String[] solution(String[] files) {
      List<strnum> answer = new ArrayList<>();
      for(int i=0; i<files.length; ++i)
      {
          answer.add(new strnum(files[i],i));
      }
      Collections.sort(answer);
      String[] result= new String[files.length];
      for(int i=0; i<result.length; ++i)
      {
        result[i]=answer.get(i).str;
      }
      return result;
  }
}